File: src/IRC/Message.hs
module IRC.Message
( Result(..)
, Message(..)
, parse
) where

import IRC.IRC
import Data.List.Split (splitOn)

type Reason = String

data Result = Success Username | Failure Reason

data Message =
    Message Channel Username String |
    Join Channel Username |
    Leave Channel Username |
    ModeratorJoin Channel Username |
    ModeratorLeave Channel Username |
    Ping String |
    Login Result

-- Safe version of !!
(!!!) :: [a] -> Int -> Maybe a
(!!!) list idx
    | idx >= length list = Nothing
    | otherwise          = Just $ list !! idx

getCode :: [String] -> Maybe String
getCode ("PING":_) = Just "PING"
getCode (_:code:_) = Just code
getCode _          = Nothing

parse :: String -> Maybe Message
parse str =
    code >>= \c -> case c of
        "PRIVMSG" -> Just $ Message channel username (tail . dropWhile (/=':') . tail $ str)
        "JOIN"    -> Just $ Join channel username
        "PART"    -> Just $ Leave channel username
        "MODE"    ->
            case sections !! 3 of
                '+':_ -> Just $ ModeratorJoin channel (sections !! 4)
                '-':_ -> Just $ ModeratorLeave channel (sections !! 4)
                _     -> Nothing
        "PING"    -> Just $ Ping $ drop (length "PING ") str
        "004"     -> Just $ Login (Success $ sections !! 2)
        "NOTICE"  -> Just $ Login (Failure $ splitOn " :" str !! 1)
        _         -> Nothing
    where
        sections = words str
        code = getCode sections
        username = maybe "ERROR" (tail . takeWhile (/='!')) (sections !!! 0)
        channel = maybe "ERROR" tail (sections !!! 2)

File: src/irc/message.rs
use ::util::string::StringUtil;

fn read_message_code(string: &str) -> Option<String> {
    // TODO: Replace with slice pattern syntax when it becomes stable
    let mut words = string.splitn(3, ' ');

    if let Some("PING") = words.next() {
        Some("PING".to_string())
    } else if let Some(code) = words.next() {
        Some(code.to_string())
    } else {
        None
    }
}

pub type Code   = String;
pub type Reason = String;

#[derive(Debug)]
pub enum MessageError {
    NoCode,
    UnknownCode(Code),
    Parse(Reason),
}

impl From<String> for MessageError {
    fn from(err: String) -> MessageError {
        MessageError::Parse(err)
    }
}

pub type Reply    = String;
pub type Contents = String;
pub type Nick     = String;

#[derive(Debug)]
pub enum Message {
    Message { nick: String, channel: String, msg: String},
    Joined  { nick: String, channel: String },
    Left    { nick: String, channel: String },
    Viewers { channel: String, viewers: Vec<Nick> },
    Ping(Reply),
    Notice(Contents),
}

impl Message {
    pub fn parse(string: &str) -> Result<Message, MessageError> {
        let get_nick = || string.between(":", "!");
        let code = read_message_code(string).ok_or(MessageError::NoCode)?;

        use self::Message::*;
        
        match code.as_str() {
            "PRIVMSG" => Ok(
                Message {
                    nick:    get_nick()?,
                    channel: string.between("#", " :")?,
                    msg:     string.after(" :")?,
                }
            ),
            "JOIN" => Ok(
                Joined {
                    nick:    get_nick()?,
                    channel: string.after(" #")?,
                }
            ),
            "PART" => Ok(
                Left {
                    nick:    get_nick()?,
                    channel: string.after(" #")?,
                }
            ),
            // List of initial channel viewers
            "353" => Ok(
                Viewers {
                    channel: string.between("#", " :")?,
                    viewers:
                        string
                        .after(" :")?
                        .split(' ')
                        .map(String::from)
                        .collect(),
                }
            ),
            "PING"   => Ok(Ping(string.after("PING ")?)),
            "NOTICE" => Ok(Notice(string.after(" :")?)),
            _        => Err(MessageError::UnknownCode(code)),
        }
    }
}